src/util/sockopt.c

```diff
@@ -194,3 +204,30 @@ int sockopt_get_peergroups(int fd, Log *log, uid_t uid, gid_t primary_gid, gid_t
 
         return 0;
 }
+
+int sockopt_get_peerpidfd(int fd, int *ret_pidfd) {
+        socklen_t socklen = sizeof(int);
+        int r, pidfd = -1;
+
+        c_assert(fd >= 0);
+        c_assert(ret_pidfd);
+
+        /*
+         * New option from kernel 6.5. We'll return this with credentials,
+         * so that access policies that are resistent from PID reuse attacks
+         * can be implemented.
+         */
+
+        r = getsockopt(fd, SOL_SOCKET, SO_PEERPIDFD, &pidfd, &socklen);
+        if (r < 0 && errno != ENOPROTOOPT)
+                return error_origin(-errno);
+
+        /* Either the option is not supported or the process already disappeared,
+         * so don't return an error but invalidate the caller's pidfd. */
```

Comment:
If I understood @dvdhrm correctly, "the process already disappeared" can't happen. We will always get a pidfd regardless. The logic is correct regardless though, so if this is unclear don't block on me bike shedding the comments ;)

Reply:
ah actually no, that's not an errno, that's a return value from the getsockopt syscall. Confusing...

Reply:
From @dvdhrm's previous comment:
> The kernel returns EINVAL if the pid has no process-group-leader attached. ESRCH is returned if SO_PEERPIDFD is called on sockets other than UDS, which have no sk_peer_pid.

So I think this case does not matter (would be a hard failure indeed).

Reply:
> ah actually no, that's not an errno, that's a return value from the getsockopt syscall. Confusing...

I just got myself confused up, never mind